python/tk_framework_desktopserver/shotgun/api_v2.py

```diff
@@ -0,0 +1,836 @@
+    ]
+    SYNCHRONOUS_METHODS = [
+        "get_actions",
+    ]
+
+    # Stores data persistently per wss connection.
+    WSS_KEY_CACHE = dict()
+    DATABASE_FORMAT_VERSION = 1
+
+    def __init__(self, host, process_manager, wss_key):
+        """
+        API Constructor.
+        Keep initialization pretty fast as it is created on every message.
+
+        :param host: Host interface to communicate with. Abstracts the client.
+        :param process_manager: Process Manager to use to interact with os processes.
+        :param str wss_key: The WSS connection's unique key.
+        """
+        self._host = host
+        self._engine = sgtk.platform.current_engine()
+        self._wss_key = wss_key
+        self._logger = sgtk.platform.get_logger("api-v2")
+
+        if self._wss_key not in self.WSS_KEY_CACHE:
+            self.WSS_KEY_CACHE[self._wss_key] = dict()
+
+        # Cache path on disk.
+        self._cache_path = os.path.join(
+            self._engine.cache_location, 
+            "shotgun_engine_commands_v%s.sqlite" % self.DATABASE_FORMAT_VERSION,
+        )
+
+    ###########################################################################
+    # Properties
+
+    @property
+    def logger(self):
+        """
+        The associated engine's logger.
+        """
+        return self._logger
+
+    @property
+    def host(self):
+        """
+        The host associated with the currnt RPC transaction.
+        """
+        return self._host
+
+    ###########################################################################
+    # Public methods
+
+    def execute_action(self, data):
+        """
+        Executes the engine command associated with the action triggered
+        in the client.
+
+        :param dict data: The payload from the client.
+        """
+        if data["name"] == "__clone_pc":
+            self.logger.debug("Clone configuration command received.")
+            try:
+                self._clone_configuration(data)
+            except Exception as e:
+                self.host.reply(
+                    dict(
+                        retcode=constants.COMMAND_FAILED,
+                        err=str(e),
+                        out=str(e),
+                    ),
+                )
+                raise
+            else:
+                self.logger.debug("Clone configuration successful.")
+                self.host.reply(
+                    dict(
+                        retcode=constants.COMMAND_SUCCEEDED,
+                        err="",
+                        out="",
+                    ),
+                )
+                return
+
+        project_entity, entities = self._get_entities_from_payload(data)
+
+        # It shouldn't ever happen at this stage of the workflow, but we'll
+        # check to make sure that we can get an environment for the type of
+        # we're dealing with here.
+        env_name = self.__pick_environment(project_entity, entities[0])
```

Comment:
What's the expected flow when the client requests actions for a HumanUser entity? Do we through through the entire caching process, detect it there, and store an empty action list for that entity type? It seems like a log of overhead to me.